connect/runtime/src/main/java/org/apache/kafka/connect/runtime/StateTracker.java

```diff
@@ -0,0 +1,107 @@
+
+/**
+ * Utility class that tracks the current state and the duration of time spent in each state.
+ */
+public class StateTracker {
+    private volatile State currentState;
+    private volatile long lastStateChangeTimeMs = 0L;
+    private volatile long unassignedTotalTimeMs = 0L;
+    private volatile long runningTotalTimeMs = 0L;
+    private volatile long pausedTotalTimeMs = 0L;
+    private volatile long failedTotalTimeMs = 0L;
+    private volatile long destroyedTotalTimeMs = 0L;
+
+    /**
+     * Change the current state.
+     *
+     * @param newState the current state; may not be null
+     * @param now      the current time in milliseconds
+     */
+    public void changeState(State newState, long now) {
+        if (currentState != newState) {
+            if (currentState != null) {
+                long duration = now - lastStateChangeTimeMs;
+                switch (currentState) {
+                    case UNASSIGNED:
+                        unassignedTotalTimeMs += duration;
+                        break;
+                    case RUNNING:
+                        runningTotalTimeMs += duration;
+                        break;
+                    case PAUSED:
+                        pausedTotalTimeMs += duration;
+                        break;
+                    case FAILED:
+                        failedTotalTimeMs += duration;
+                        break;
+                    case DESTROYED:
+                        destroyedTotalTimeMs += duration;
+                        break;
+                }
+            }
+            currentState = newState;
+            lastStateChangeTimeMs = now;
+        }
+    }
+
+    /**
+     * Calculate the ratio of time spent in the specified state.
+     *
+     * @param ratioState the state for which the ratio is to be calculated; may not be null
+     * @param now        the current time in milliseconds
+     * @return the ratio of time spent in the specified state to the time spent in all states
+     */
+    public double durationRatio(State ratioState, long now) {
+        if (currentState == null) return 0.0d;
+        long durationCurrent = now - lastStateChangeTimeMs; // since last state change
+        long durationDesired = ratioState == currentState ? durationCurrent : 0L;
+        switch (ratioState) {
+            case UNASSIGNED:
+                durationDesired += unassignedTotalTimeMs;
+                break;
+            case RUNNING:
+                durationDesired += runningTotalTimeMs;
+                break;
+            case PAUSED:
+                durationDesired += pausedTotalTimeMs;
+                break;
+            case FAILED:
+                durationDesired += failedTotalTimeMs;
+                break;
+            case DESTROYED:
+                durationDesired += destroyedTotalTimeMs;
+                break;
+        }
+        long total = durationCurrent + unassignedTotalTimeMs + runningTotalTimeMs + pausedTotalTimeMs +
```

Comment:
Either this is not atomic or I'm missing something.

Reply:
ack.